File: betterrandom/src/test/java/io/github/pr0methean/betterrandom/prng/RandomWrapperAesCounterRandomTest.java
package io.github.pr0methean.betterrandom.prng;

import com.google.common.collect.ImmutableList;
import io.github.pr0methean.betterrandom.seed.SeedException;
import java.util.Random;
import org.testng.annotations.Test;

@Test(testName = "RandomWrapper:AesCounterRandom")
public class RandomWrapperAesCounterRandomTest extends AesCounterRandomTest {

  private static final NamedFunction<Random, Double> SET_WRAPPED = new NamedFunction<>(random -> {
    ((RandomWrapper) random).setWrapped(new AesCounterRandom());
    return 0.0;
  }, "setWrapped");

  public RandomWrapperAesCounterRandomTest() {
    super(16);
  }

  @Override @Test public void testThreadSafetySetSeed() {
    testThreadSafetyVsCrashesOnly(30,
        ImmutableList.of(NEXT_LONG, NEXT_INT, NEXT_DOUBLE, NEXT_GAUSSIAN, SET_SEED, SET_WRAPPED));
  }

  @Override @Test(enabled = false) public void testAdvanceForward() {
    // No-op: RandomWrapper isn't seekable
  }

  @Override @Test(enabled = false) public void testAdvanceBackward() {
    // No-op: RandomWrapper isn't seekable
  }

  @Override @Test(enabled = false) public void testAdvanceZero() {
    // No-op: RandomWrapper isn't seekable
  }

  @Override protected Class<? extends BaseRandom> getClassUnderTest() {
    return RandomWrapper.class;
  }

  @Override @Test(enabled = false) public void testAllPublicConstructors() {
    // No-op: redundant to super insofar as it works.
  }

  @Override protected RandomWrapper createRng() throws SeedException {
    return new RandomWrapper(new AesCounterRandom(getTestSeedGenerator()));
  }

  @Override protected RandomWrapper createRng(final byte[] seed) throws SeedException {
    return new RandomWrapper(new AesCounterRandom(seed));
  }
}

File: betterrandom/src/test/java/io/github/pr0methean/betterrandom/prng/RandomWrapperAesCounterRandomTest.kt
package io.github.pr0methean.betterrandom.prng

import com.google.common.collect.ImmutableList
import io.github.pr0methean.betterrandom.seed.SeedException
import java.util.Random
import org.testng.annotations.Factory
import org.testng.annotations.Test

@Test(testName = "RandomWrapper:AesCounterRandom")
class RandomWrapperAesCounterRandomTest : AesCounterRandomTest() {

    protected override val classUnderTest: Class<out BaseRandom>
        get() = RandomWrapper::class.java

    init {
        seedSizeBytes = 16
    }

    @Test
    override fun testThreadSafetySetSeed() {
        testThreadSafetyVsCrashesOnly(30,
                ImmutableList.of(BaseRandomTest.NEXT_LONG, BaseRandomTest.NEXT_INT, BaseRandomTest.NEXT_DOUBLE, BaseRandomTest.NEXT_GAUSSIAN, BaseRandomTest.SET_SEED, SET_WRAPPED))
    }

    @Test(enabled = false)
    override fun testAdvanceForward() {
        // No-op: RandomWrapper isn't seekable
    }

    @Test(enabled = false)
    override fun testAdvanceBackward() {
        // No-op: RandomWrapper isn't seekable
    }

    @Test(enabled = false)
    override fun testAdvanceZero() {
        // No-op: RandomWrapper isn't seekable
    }

    @Test(enabled = false)
    override fun testAllPublicConstructors() {
        // No-op: redundant to super insofar as it works.
    }

    @Throws(SeedException::class)
    override fun createRng(): RandomWrapper {
        return RandomWrapper(AesCounterRandom(testSeedGenerator))
    }

    @Throws(SeedException::class)
    override fun createRng(seed: ByteArray?): RandomWrapper {
        return RandomWrapper(AesCounterRandom(seed))
    }

    companion object {

        private val SET_WRAPPED = BaseRandomTest.NamedFunction<Random, Double>({ random ->
            (random as RandomWrapper).wrapped = AesCounterRandom()
            0.0
        }, "setWrapped")
    }
}
